src/configuration.js

```diff
@@ -0,0 +1,252 @@
+     *
+     * @param {Record<string, number | string | boolean>} options
+     */
+    constructor(options = {}) {
+        this.options = options;
+
+        if (!this.get('localStorageDir') && !this.get('token')) {
+            const dir = join(process.cwd(), './apify_storage');
+            this.set('localStorageDir', dir);
+            // eslint-disable-next-line max-len
+            log.warning(`Neither ${ENV_VARS.LOCAL_STORAGE_DIR} nor ${ENV_VARS.TOKEN} environment variable is set, defaulting to ${ENV_VARS.LOCAL_STORAGE_DIR}="${dir}"`);
+        }
+    }
+
+    /**
+     * Returns configured value. First checks the environment variables, then provided configuration,
+     * fallbacks to the `defaultValue` argument if provided, otherwise uses the default value as described
+     * in the above section.
+     *
+     * @param {string} key
+     * @param {string | number | boolean} [defaultValue]
+     * @return {string | number | boolean}
+     */
+    get(key, defaultValue) {
+        // prefer env vars
+        const envKey = Configuration.ENV_MAP_REVERSED[key] ?? key;
+        const envValue = process.env[envKey] ?? process.env[ENV_VARS[envKey] ?? process.env[`APIFY_${envKey}`]];
+
+        if (envValue != null) {
+            return this._cast(key, envValue, true);
+        }
+
+        // check instance level options
+        if (key in this.options) {
+            return this.options[key];
+        }
+
+        // fallback to defaults
+        return defaultValue ?? Configuration.DEFAULTS[key] ?? envValue;
+    }
+
+    /**
+     * @param {string} key
+     * @param {number | string | boolean} value
+     * @param {boolean} [env=false]
+     * @return {boolean}
+     * @private
+     */
+    _cast(key, value, env = false) {
+        if (key in Configuration.INTEGER_VARS) {
+            return +value;
+        }
+
+        if (key in Configuration.BOOLEAN_VARS) {
+            if (env) {
+                // 0, false and empty string are considered falsy values
+                return !['0', 'false', ''].includes(value.toLowerCase());
+            }
+
+            return !!value;
+        }
+
+        return value;
+    }
+
+    /**
+     * Sets value for given option. Only affects this `Configuration` instance, the value will not be propagated down to the env var.
+     * To reset a value, we can omit the `value` argument or pass `undefined` there.
+     *
+     * @param {string} key
+     * @param {string | number | boolean} [value]
+     */
+    set(key, value) {
+        this.options[key] = value;
+    }
+
+    /**
+     * Creates an instance of ApifyClient using options as defined in the environment variables or in this `Configuration` instance.
+     *
+     * @param {object} [options]
+     * @param {string} [options.token]
+     * @param {string} [options.maxRetries]
+     * @param {string} [options.minDelayBetweenRetriesMillis]
+     * @return {ApifyClient}
+     * @internal
+     */
+    createClient(options = {}) {
```

Comment:
As mentioned before, `createClient` and `createLocalStorage` look a bit out of place here. What's the plan for them?

Reply:
I just noticed that this does not memoize the client instance. We need to have only a single client instance because the `AutoscaledPool` measures API performance based on information from that one client. If each storage has its own client then the pool would not be able to scale based on API rate limits because it can't read that info from the clients.

Reply:
Can you point me to the problematic part? As I really don't understand why this should matter for the client, it should be stateless, apparently it is not? Why? I can understand the storage local, if it creates db connections, we should have single instance - but the end goal should be to have single instance of the config, not to enforce this on the storage local level. 

As mentioned above, we can store the instance in the `Configuration` object for now to have singletons, if it's really needed.

Also for me it really is not that much out of place, the `Configuration` object will be kind of a DI container (there should be just one instance of it). If something was out of place, it was the previous implementation - you require `utils` (which is automaticly done when required the package I guess) and suddenly things get instantiated and db connections created (and even if you don't use the storage local, it will get instantiated)...

Reply:
The problem is that the client is not entirely stateless. It measures error statistics, especially rate limits and we use the rate limits in the `AutoscaledPool` to scale the actors. The pool uses 4 metrics to determine whether it can add more concurrency. CPU, memory, event loop and API rate limits. That's because it does not make sense to scale up the crawlers if you for example can't save data fast enough or pull requests from the queue fast enough, even if you have extra CPU power.

It's not ideal, but it is the current situation. If we have multiple clients, the rate limit scaling will not work, because the `AutoscaledPool` (actually the `Snapshotter`) relies on periodically reading those rate limit statistics to determine the API performance.

Reply:
It's this function: https://github.com/apify/apify-js/blob/935720567985a5e01a6f4d501f21eec2497b0354/src/autoscaling/snapshotter.js#L424

Reply:
And yeah, I agree that the previous implementation was much worse 😉